File: benchmarks/pandas/bench_nat_sort.py
"""Benchmark: natural sort using natsort library (equivalent to natSorted/natArgSort)."""
import json, time

SIZE = 10_000
WARMUP = 5
ITERATIONS = 50

data = [f"item{i % 1000}_v{i % 10}" for i in range(SIZE)]

try:
    from natsort import natsorted, index_natsorted
    def run():
        natsorted(data)
        index_natsorted(data)
except ImportError:
    import re
    def nat_key(s):
        return [int(c) if c.isdigit() else c.lower() for c in re.split(r'(\d+)', s)]
    def run():
        sorted(data, key=nat_key)
        sorted(range(len(data)), key=lambda i: nat_key(data[i]))

for _ in range(WARMUP):
    run()

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    run()
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({"function": "nat_sort", "mean_ms": round(total_ms / ITERATIONS, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))

File: benchmarks/tsb/bench_nat_sort.ts
/**
 * Benchmark: natSorted / natCompare / natArgSort — natural sort.
 * Outputs JSON: {"function": "nat_sort", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { natSorted, natCompare, natArgSort } from "../../src/index.ts";

const SIZE = 10_000;
const WARMUP = 5;
const ITERATIONS = 50;

const data = Array.from({ length: SIZE }, (_, i) => `item${i % 1000}_v${i % 10}`);

for (let i = 0; i < WARMUP; i++) {
  natSorted(data);
  natArgSort(data);
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const start = performance.now();
  natSorted(data);
  natArgSort(data);
  times.push(performance.now() - start);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "nat_sort",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);
